File: patterns/java/src/refactoring_guru/patterns/observer/example/EventManager.java
package refactoring_guru.patterns.observer.example;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EventManager {
    Map<String, List<EventListener>> listeners = new HashMap<>();
    List<EventListener> open = new ArrayList<>();
    List<EventListener> save = new ArrayList<>();

    public EventManager() {
        this.listeners.put("open", open);
        this.listeners.put("save", save);
    }

    public void subscribe(String eventType, EventListener listener) {
        List<EventListener> users = listeners.get(eventType);
        users.add(listener);
    }

    public void unsubscribe(String eventType, EventListener listener) {
        List<EventListener> users = listeners.get(eventType);
        int index = users.indexOf(listener);
        users.remove(index);
    }

    public void execute(String eventType, File file) {
        List<EventListener> users = listeners.get(eventType);
        for (EventListener listener : users) {
            listener.update(file);
        }
    }
}

File: patterns/cs/src/RefactoringGuru/Observer/Example/Publisher/EventManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using RefactoringGuru.Observer.Example.Listeners;

namespace RefactoringGuru.Observer.Example.Publisher
{
    class EventManager
    {
        Dictionary<String, List<EventListener>> listeners = new Dictionary<string, List<EventListener>>();

        public EventManager(String[] operations)
        {
            foreach (string str in operations)
            {
                this.listeners.Add(str, new List<EventListener>());
            }
        }

        public void Subscribe(String eventType, EventListener listener)
        {
            List<EventListener> users = listeners[eventType];
            users.Add(listener);
        }

        public void Unsubscribe(String eventType, EventListener listener)
        {
            List<EventListener> users = listeners[eventType];
            int index = users.IndexOf(listener);
            users.RemoveAt(index);
        }

        public void Notify(String eventType, FileInfo file)
        {
            IList<EventListener> users = listeners[eventType];
            foreach (var listener in users)
            {
                listener.Update(eventType, file);
            }
        }
    }
}
